### ISSUE TITLE
exhaustive type check for recursive type compositions

### ISSUE THREAD

**Issue by romansky** [microsoft/TypeScript#36182]
**TypeScript Version:**  3.7.x-dev.201xxxxx

**Search Terms:**
recursive types, exhaustive checking
**Code**

```ts
const tc = <T, PT>(
    cIsT: (c: T | PT) => c is T,
    value: string,
    pt: (x: PT) => string | undefined = (_) => undefined)
    : ((resolve: PT | T) => string | undefined) => (resolve) => !!cIsT(resolve) ? value : pt(resolve);

type types = string | number;

const isString = (input: string | any): input is string => typeof input === 'string';
const isNumber = (input: number | any): input is number => typeof input === 'number';

// below should compile fine
const resolver: (pt: string | number) => (string | undefined) = tc(isString, 'str', tc(isNumber, 'numnum'));

// below should show error as "tc" implementation does not include `boolean` type
const resolver: (pt: string | number | boolean) => (string | undefined) = tc(isString, 'str', tc(isNumber, 'numnum'));

// usage example
console.log(resolver("asdf"));
```

**Expected behavior:**
Compilation error, `tc` is not exhaustive, missing boolean
**Actual behavior:**
runs..
**Playground Link:** 
https://www.typescriptlang.org/play/?target=6&ssl=13&ssc=1&pln=13&pc=31#code/MYewdgzgLgBFwwLwwDwBUA0MAKaB8AFAFAykzACSEaAXDAcHWjAD45oCUSe5MAlhBiYSZAG4BDADYBXAKZ1oAJz5gA5hhGkADlDoEAHnVxdEPJStWsY0sABNZAMxWzbSegH0TPG-adgXHJowegSKshAgkqLy7Fac3DDmalY+js62XvRhEVGymQCE+ZTUoeGR0VwA-DASMrLBMDqlORUA3EREUACeWvXdvYLISZZsYNIAtgBGsortRKCQsAIAylDKycgEKlrSuolrFlbiYF0cdNu7-ILDCf2yIA78YDuwiG8wAOTDH+0L0FcAOQm00Ubi2z12dDGUxmRxOZyeLyuMGhINuPXujwur3eH1RMx+HT+sGy5Rmeh0CgOyVGwNhbEmIEismOmQINzYqT8ATc8C2EFW63UnyUHywfIEQJhiiweIm0I+HA4v3AOVkADpJCBVM0yYoCAAicQQWwOA1K1pAA
**Related Issues:**

**Comment by RyanCavanaugh** [comment]
@romansky what's the simplest version of this that demonstrates the problem you're describing?

**Comment by romansky** [comment]
I was hoping this was a simple enough version.. 
The type signature below requires that `pt` is a union type between `string | number | boolean`, but on right side I'm only implementing for `number` and `string`, should the compiler be able to determine that the `tc(isString, 'str', tc(isNumber, 'numnum'))` expression only boils down to `(pt: string | number) => (string | undefined)` and that `boolean` is missing? 

```
const resolver: (pt: string | number | boolean) => (string | undefined) = tc(isString, 'str', tc(isNumber, 'numnum'));
```

**Comment by nmain100** [comment]
If I'm understanding the intent of the example correctly, I don't think there's anything wrong here.  The types are behaving as expected, they just don't guarantee what you think is being guaranteed.

Here's a smaller, simpler repro that exhibits the same behavior:

```ts
function f(a: string | any) {
    if (typeof a === "string") {
        console.log("It's a string");
    }
}

const g: (a: string | number | any) => void = f;
```

You'd like for g to print `"It's a number"` when passed a number, but it doesn't, and typescript doesn't reject it!  The rest of the example is mostly window dressing, and it boils down to the `| any` on the parameter types of your functions that lets them accept, well, anything.

More specifically, the `PT` type argument to `tc` is inferred as `any` in all cases in your example, which gives the the argument type of the returned resolver function as `any`.

**Comment by romansky** [comment]
@nmain100 your example is not relevant..

@RyanCavanaugh here's a simpler example

```
const tc = <T, PT>(t: (_t: T) => void, pt: (_pt: PT) => void): ((input: T | PT) => void) => { throw new Error("doesn't matter"); }

tc<string, never>((x: string) => null, (_: never) => null)
// below should not compile as second type argument needs to be string | never
tc<string, never>((x: string) => null, tc<string, never>((x: string) => null, (_: never) => null))
```

[Playground Link](https://www.typescriptlang.org/play/?target=6#code/MYewdgzgLgBFwwLwwDwBUA0MAKaB8AFFAFwwED6JMaAlEnjAG4gCWAJlgA5UXem51EDZuxqkCBFmE4BXKmhgAfHLXpNWbQQwDecABYAnEAHcYYAKamAogaMGCAIgc0A3DAC+AKE-wU0A1IA5lgWjOYGhAQAHqT+QVpmMgA2SVgUpKHhCWDJSTSeAPQFMABG5kkmMBB6IMlsZiCwoAC2nCxJ5jAAhhBV5qBg9VAAnpydXQaBMs3mYLAW5my9UCClnXFggUpm5mEGPsB+UAGbIbvhkTFVx-FqOSlYvhvBO3uXsTeb2blp5BnnBm+KRoNCAA)

**Comment by nmain100** [comment]
This is just the basics of the type system.  In your second example, the type of `tc<string, never>((x: string) => null, (_: never) => null)` is `(input: string | never) => void`, which is exactly equivalent to `(input: string) => void` and is assignable to `(input: never) => void` because functions are contravariant in their inputs.

**Comment by RyanCavanaugh** [comment]
```
// below should not compile as second type argument needs to be string | never
```
I don't see any evidence of a bug here. `string | never` is equivalent to `string`, the same way `5 + 0` is `5`.

The OP example is predicated on a function taking `string | any` which is equivalent to `any`, so assignability sourced in that function is not a surprise.
